api.py

```diff
@@ -0,0 +1,75 @@
+from functools import lru_cache, partial
+from typing import Callable, Dict, List, Optional, Tuple
+import os
+
+import flask
+
+app = flask.Flask(__name__)
+
+FILESYSTEM_ROOT = os.path.abspath(os.sep)
+DATASETS_ROOT = os.path.join(FILESYSTEM_ROOT, 'data')
+
+
+def readable(dataset_name: str) -> str:
+    """Convert name on disk to readable for the user"""
+    return dataset_name.replace('_', ' ')
+
+
+def get_datasets() -> List[Dict[str, str]]:
+    """"Get datasets available in the store"""
+    datasets = filter(os.path.isdir, os.listdir(DATASETS_ROOT))
+    return [{"name": readable(name), "value": name} for name in datasets]
+
+
+def substitute_tags(tag_map: Dict[str, str], text: str) -> str:
+    """Substitute tags from the text for corresponding values in the map"""
+    for tag in tag_map:
+        text = text.replace('"' + tag + '"', tag_map[tag])
+    return text
+
+
+Substitutor = Optional[Callable[[str], str]]
+
+
+def datasets_substitutor() -> Substitutor:
+    """Factory of datasets substitutor"""
+    datasets = get_datasets()
+    parsed = str(datasets).replace("'", '"')
+    return partial(substitute_tags, {'$DATASETS': parsed})
+
+
+SubstitutorFactory = Callable[[], Substitutor]
+Response = Tuple[str, int]
+NOT_FOUND = "", 404
+
+
+def file_from_disk(substitutor_factory: SubstitutorFactory, path: str) -> Response:
+    """Read file from disk with subsitutions and return it as HTTP response"""
+    if not os.path.exists(path):
+        return NOT_FOUND
+    with open(path) as disk_file:
+        content = disk_file.read()
+    if substitutor_factory is None:
+        return content, 200
+    substitutor = substitutor_factory()
+    substituted = substitutor(content)
+    return substituted, 200
+
+
+file_with_datasets_substitution = partial(file_from_disk, datasets_substitutor)
+unchanged_file = partial(file_from_disk, None)
+
+
+@app.route('/schema/<string:endpoint>/<string:task_name>/')
+# @lru_cache()
+def schema(endpoint:str, task_name: str) -> Response:
+    """Get static schema description"""
+    path = os.path.join('.', 'schema', endpoint, task_name + '.json')
+    return unchanged_file(path)
```

Comment:
The function seems not to be really susceptible to path traversal attacks when called from Flask’s web context, as the highest possible path is `./schema/../some.json`, which is not dangerous. However, for the sake of security, I would recommend restricting paths to `./schema` directory [this way](https://stackoverflow.com/a/45188896).

Please respond with `403 Forbidden` when outside of secure path.

Reply:
[No, it's not.](http://flask.pocoo.org/docs/0.12/quickstart/#variable-rules) That was my first thought :)

`<string:...>` quantifier restricts the argument to contain no `/`. If we wanted to allow this, we would have to use `<path:...>` quantifier.